### PR TITLE
[feat] Add optimized clusters TopK kernel and DeepSeekV3.2 sparse indexer fused kernel

### PR DESCRIPTION
**Pull request by Aalanli**
This PR specifically targets low-latency decode stage for deepseek V3.2's sparse indexer. It includes an optimized kernel implementation for the entire indexer.

## Comparison against baseline: deep gemm + flashinfer topK
This is the entire indexer where the deepgemmm_ref uses deepgemm's `fp8_paged_mqa_logits` with flashinfer's topK.
```
> python benchmarks/bench_dsv3_sparse_indexer.py --pdl

====================================================================================================
dsv3_sparse_indexer: DeepSeek v3 sparse-attention end-to-end (k=2048, fp8) PDL=on
  mqa_fused:     FlashInfer fused logits + histogram top-K
  mqa_non_fused: FlashInfer separate logits + radix top-K
  deepgemm_ref:  deep_gemm fp8_paged_mqa_logits + flashinfer top_k
====================================================================================================
 batch    seq_len |    mqa_fused   mqa_non_fused   deepgemm_ref    speedup
----------------------------------------------------------------------------------
     1       1024 |       6.82us          6.46us        11.14us      1.63x
     1       4096 |      15.87us         16.16us        16.22us      1.02x
     1       8192 |      15.78us         16.10us        18.88us      1.20x
     1      32768 |      17.20us         17.73us        29.54us      1.72x
     1      40960 |      18.66us         18.82us        39.49us      2.12x
     2       1024 |       7.71us          6.66us        11.55us      1.50x
     2       4096 |      15.65us         16.10us        16.13us      1.03x
     2       8192 |      16.26us         16.45us        19.52us      1.20x
     2      32768 |      18.14us         18.88us        30.08us      1.66x
     2      40960 |      19.65us         20.67us        41.66us      2.12x
     4       1024 |       8.13us          7.07us        11.60us      1.43x
     4       4096 |      15.71us         16.13us        17.46us      1.11x
     4       8192 |      16.90us         17.15us        19.82us      1.17x
     4      32768 |      20.67us         20.96us        31.71us      1.53x
     4      40960 |      21.12us         22.02us        37.25us      1.76x
     8       1024 |       8.22us          7.10us        12.22us      1.49x
     8       4096 |      16.77us         17.12us        18.11us      1.08x
     8       8192 |      17.50us         17.86us        20.99us      1.20x
     8      32768 |      24.03us         24.66us        35.26us      1.47x
     8      40960 |      25.73us         26.54us        40.73us      1.58x
    32       1024 |       8.77us          8.19us        13.06us      1.49x
    32       4096 |      21.86us         22.40us        21.15us      0.97x
    32       8192 |      25.14us         25.76us        25.74us      1.02x
    32      32768 |      46.99us         47.86us        58.98us      1.26x
    32      40960 |      54.70us         56.10us        66.43us      1.21x
    64       1024 |      10.82us          9.82us        14.24us      1.32x
    64       4096 |      34.85us         36.82us        24.56us      0.70x
    64       8192 |      42.69us         44.96us        34.03us      0.80x
    64      32768 |      85.10us         87.90us        84.48us      0.99x
    64      40960 |      98.69us        101.58us        97.10us      0.98x
```

## Comparison against VLLM's indexer
image

### Epilogue fusion does not affect gemm performance
image

### TopK gets speedup
image

### CODE REVIEW

**Comment by kahyunnam** [comment]
@Aalanli what compute capability did you optimize for / on what gpu did you collect these benchmarking results?

**Comment by Aalanli** [comment]
Hi @kahyunnam , I benchmarked on a bare-metal B200 node, for sm100a. Note that the topK works for both hopper and blackwell but the gemm kernel uses tcgen05 instructions so only works for blackwell. Most of the speedup I observed comes from the topK, but the histogram fusion and PDL is a slightly added boost. 

This PR is mostly to get the comparisons going, the code still needs a lot of cleanup if we want to merge this.

**Comment by Aalanli** [comment]
Update, I implemented another topk algorithm that is exact, with only minor overhead. The gist of it is that depending on the distribution of logits, there might be greater than the shared cache number of logits in the threshold bin, thus in the inexact version we drop them, but for the exact version we overflow to a global cache. This does not have too much overhead for the normal distribution. 

```
====================================================================================================
dsv3_topk: DeepSeek v3 sparse-attention top-K shapes (k=2048, float32)
  fast_topk_clusters:       histogram-driven radix top-K (variable-length aware)
  fast_topk_clusters_exact: overflow-safe variant (always returns exact top-2048)
  flashinfer top_k:         generic radix top-K (padded input)
====================================================================================================
 batch    seq_len |   fast_topk_clusters        exact   overhead   flashinfer_top_k    Speedup
----------------------------------------------------------------------------------------------------
     1       1024 |               1.79us       1.84us      1.03x             7.36us      4.11x
     1       4096 |              10.30us      10.53us      1.02x            12.43us      1.21x
     1       8192 |              10.37us      10.66us      1.03x            14.82us      1.43x
     1      32768 |              11.42us      12.06us      1.06x            25.04us      2.19x
     1      40960 |              12.54us      13.22us      1.05x            37.28us      2.97x
     2       1024 |               1.79us       1.79us      1.00x             7.94us      4.43x
     2       4096 |              10.37us      10.62us      1.02x            12.48us      1.20x
     2       8192 |              10.40us      10.66us      1.02x            14.91us      1.43x
     2      32768 |              11.52us      12.21us      1.06x            27.46us      2.38x
     2      40960 |              12.58us      13.18us      1.05x            37.86us      3.01x
     4       1024 |               1.82us       1.82us      1.00x             7.94us      4.35x
     4       4096 |              10.34us      10.53us      1.02x            11.30us      1.09x
     4       8192 |              10.43us      10.75us      1.03x            15.33us      1.47x
     4      32768 |              11.58us      12.26us      1.06x            27.65us      2.39x
     4      40960 |              12.61us      13.25us      1.05x            29.57us      2.35x
    32       1024 |               1.92us       1.92us      1.00x             8.19us      4.27x
    32       4096 |              12.99us      13.28us      1.02x            11.57us      0.89x
    32       8192 |              13.38us      13.66us      1.02x            17.54us      1.31x
    32      32768 |              15.20us      15.87us      1.04x            29.82us      1.96x
    32      40960 |              16.82us      17.41us      1.04x            29.82us      1.77x
    64       1024 |               1.92us       1.95us      1.02x             8.35us      4.35x
    64       4096 |              12.05us      12.29us      1.02x            13.47us      1.12x
    64       8192 |              13.62us      13.82us      1.02x            17.60us      1.29x
    64      32768 |              17.34us      17.82us      1.03x            29.89us      1.72x
    64      40960 |              19.87us      20.38us      1.03x            30.11us      1.52x
```

image

I've also modified the number of clusters depending on batch size, for higher batch, it's better to use a smaller cluster size.

File: csrc/dsv3_ops/fast_topk_clusters.cu [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+        }
+      }
+
+      if (bin == threshold_bin) {
+        int cached_offset = atomicAdd(&shared_num_cached_count[0], 1);
+        if (cached_offset < num_cached) {
+          s_cached_indices[cached_offset] = i;
+          s_cached_logit_bits[cached_offset] = bits;
+          atomicAdd(shared_hist[1] + (bits >> 16 & 0xff), 1);
+        }
```
[the remark this answers is not in the document]
Comment (Aalanli) [reply csrc/dsv3_ops/fast_topk_clusters.cu:141]: This is by design, the exact version trades off performance and eliminates this concern.

File: csrc/dsv3_ops/fast_topk_clusters_exact.cu [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+  DISPATCH_TOPK_EXACT(1, true);
+  DISPATCH_TOPK_EXACT(2, true);
+  DISPATCH_TOPK_EXACT(4, true);
+  DISPATCH_TOPK_EXACT(8, true);
+
+  if (kernel == nullptr) {
+    num_clusters = 1;
+    pdl_enabled = false;
+    DISPATCH_TOPK_EXACT(1, false);
+  }
```
[the remark this answers is not in the document]
Comment (Aalanli) [reply csrc/dsv3_ops/fast_topk_clusters_exact.cu:410]: I think it's fine.

**Review by bkryu — COMMENTED** [review]
Hi @Aalanli, thanks for contributing this. Just so I understand:

This PR adds three public APIs:
* `flashinfer.mqa_topk_indexer()` --> 2 kernel API where
  * Kernel 1: Matmul -> ReLU -> Weighted Reduction -> Scale -> Histogram & logits
  * Kernel 2: Histogram -> topK
* `flashinfer.mqa_topk_indexer_non_fused()` --> 2 kernel API where
  *  Kernel 1: Matmul -> ReLU -> Weighted Reduction -> Scale -> logits but no histogram
  * Kernel 2:  Scan logits to build historgram -> topK
* `flashinfer.get_mqa_metadata()` helper function that generates metadata for the two APIs above.

Am I understanding this correctly? If so, this PR contributes something beyond a simple top-K kernel -- it sounds like `mqa_topk_indexer` is the full lightning indexer specialized in DSv3.2. For these types of specialized APIs, we can place them in the `dsv3_ops` [directory](https://github.com/flashinfer-ai/flashinfer/tree/main/flashinfer/dsv3_ops)

Do you mind moving the files to over there?

**Comment by Aalanli** [comment]
Hi @bkryu, I've updated the front facing code and renamed everything to be more clear. Now the only public facing apis are 
- `dsv3_topk_indexer`: computes the gemm+weight+sum+topk fusion 
- `get_indexer_metadata`: computes the sm load balancing metadata for the indexer. This can be computed once in the forward pass, so amortized. 

By default I use the exact topk since I don't observe much perf degradation, but there's an argument that let's users decide.

**Comment by nvpohanh** [comment]
cc @nvjullin

**Review by kahyunnam — COMMENTED** [review]
Can we add the new fused APIs to the documentation (see: https://github.com/flashinfer-ai/flashinfer/blob/main/docs/api/topk.rst) so that users know about it?

File: flashinfer/dsv3_ops/indexer.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        - logits = (relu(q[bi] @ dequant_kvcache) * weights[bi][:, None]).sum(0) # [seq_len]
+        - topk(logits, k = 2048)
+
+    Args:
+        q:           [batch, 64, 128] fp8/uint8 CUDA tensor.
+        k_cache:     [num_pages, 64, 1, 132] fp8/uint8 CUDA tensor.
+        weights:     [batch, 64] float32 CUDA tensor.
+        seq_lens:    [batch] int32 CUDA tensor.
+        block_table: [batch, max_num_pages] int32 CUDA tensor.
+        sm_map:      Optional [num_sms, 4] int32 tensor from get_indexer_metadata().
```
Comment (kahyunnam) [inline flashinfer/dsv3_ops/indexer.py:181]: Thanks for the clear documentation on the expected shapes; it might also be good practice to have an assert check for these at the API level

Comment (Aalanli) [reply flashinfer/dsv3_ops/indexer.py:181]: Yes, I will do that.

**Comment by wzhao18** [comment]
@Aalanli Would this work with Deepseek V4? Also, do you have a branch in vllm that integrates the kernel?

**Comment by Aalanli** [comment]
Hi @wzhao18, DSV4's kernel is different, but seems like a similar epilogue fusion as done here is possible. I have a hacky branch that jits the kernel, not one that uses flashinfer. 

For now I think it's best to use DeepGEMM's kernel as experiments show that most of the perf uplift comes from the topk which is already merged. There are several performance enhancements that could be done to this kernel that I've thought of:
- using persistent kernel with cluster launch control
- use `tile::gather4` instruction for pages and scales loading
but unfortunately I lack access to b200s to test these ideas out.

**Comment by wzhao18** [comment]
@Aalanli Got it. I wanted to ping because if you think this work is valuable, especially given the release of DSv4 which also uses sparse attention, you should consider asking for help to continue the work or it would go stale.
